### PR TITLE
httpreq supports `Content-Type`, `Transport`, `httpClient` parameters

### PR DESCRIPTION
**Pull request by xiaoxinmiao**

### CODE REVIEW

File: httpreq/httpreq.go [hunk 1]
```diff
@@ -12,26 +11,20 @@ import (
...
-	defaultTransportPointer, ok := http.DefaultTransport.(*http.Transport)
-	if !ok {
-		panic(fmt.Sprintf("defaultRoundTripper not an *http.Transport"))
-	}
-	defaultTransport := *defaultTransportPointer
-
-	// http.DefaultMaxIdleConnsPerHost = 2
-	defaultTransport.MaxIdleConnsPerHost = DefaultMaxIdleConnsPerHost
+const (
+	dataTypeKey key = iota
```
Comment (jaehue) [inline httpreq/httpreq.go]: When use `dataTypeKey`? If we don't use it, please remove this `const` part.

File: httpreq/httpreq.go [hunk 2]
```diff
@@ -58,12 +57,33 @@ func New(method, url string, param interface{}) *HttpReq {
...
-	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Content-Type", dataType)
 	return &HttpReq{
-		req: req,
+		req:      req,
+		dataType: dataType,
 	}
 }
 
+func (r *HttpReq) setDefaultClient() *HttpReq {
```
Comment (jaehue) [inline httpreq/httpreq.go]: Maybe, we almost use `setDefaultClient()` method.
How about create `defaultClient` variable in `init` method?
Then `setDefaultClient()` method can reuse `defaultClient`.

File: httpreq/util.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	MIMEApplicationMsgpack               = "application/msgpack"
+	MIMETextHTML                         = "text/html"
+	MIMETextHTMLCharsetUTF8              = "text/html" + "; " + charsetUTF8
+	MIMETextPlain                        = "text/plain"
+	MIMETextPlainCharsetUTF8             = "text/plain" + "; " + charsetUTF8
+	MIMEMultipartForm                    = "multipart/form-data"
+	MIMEOctetStream                      = "application/octet-stream"
+)
+
+type httpData interface {
```
Comment (jaehue) [inline httpreq/util.go]: Commonly, the postfix of interface name is '-er' in Go. How about `reqFormatter`?

File: httpreq/util.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	marshal(data interface{}) ([]byte, error)
+	Head() string
```
Comment (jaehue) [inline httpreq/util.go]: Only `Head()` method is exported, but others are unexported. It is better to use same convention.

File: httpreq/httpreq.go (hunk 1 above)
Comment (relax-admin) [reply httpreq/httpreq.go]: ok，I removed it

File: httpreq/httpreq.go (hunk 2 above)
Comment (relax-admin) [reply httpreq/httpreq.go]: ok,I use init method

File: httpreq/util.go (hunk 3 above)
Comment (relax-admin) [reply httpreq/util.go]: this name is very good

File: httpreq/util.go (hunk 4 above)
Comment (relax-admin) [reply httpreq/util.go]: ok,use the same convention.

File: httpreq/httpreq_test.go [hunk 5]
```diff
@@ -6,10 +6,14 @@ import (
 	"fmt"
 	"net/http"
 	"net/http/httptest"
+	"os"
 	"testing"
+	"time"
 
 	"github.com/pangpanglabs/goutils/behaviorlog"
 
+	httpreq2 "goutils/httpreq"
```
Comment (jaehue) [inline httpreq/httpreq_test.go]: Do not use `goutils/httpreq` package.

File: httpreq/httpreq.go [hunk 6]
```diff
@@ -17,21 +16,13 @@ const DefaultMaxIdleConnsPerHost = 100
...
-	if !ok {
-		panic(fmt.Sprintf("defaultRoundTripper not an *http.Transport"))
-	}
-	defaultTransport := *defaultTransportPointer
-
-	// http.DefaultMaxIdleConnsPerHost = 2
-	defaultTransport.MaxIdleConnsPerHost = DefaultMaxIdleConnsPerHost
-
-	defaultClient = &http.Client{Transport: &defaultTransport}
+	defaultClient = &http.Client{}
```
Comment (jaehue) [inline httpreq/httpreq.go]: Setup `defaultClient.Transport` in `init` function.

File: httpreq/httpreq.go [hunk 7]
```diff
@@ -76,6 +90,18 @@ func (r *HttpReq) WithToken(token string) *HttpReq {
 	return r
 }
 
+func (r *HttpReq) WithTransport(transport *http.Transport) *HttpReq {
```
Comment (jaehue) [inline httpreq/httpreq.go]: Do not change `defaultClient`.
I think it is better `CallWithTransport` or `CallWithHttpClient` rather then `WithTransport`.

Comment (relax-admin) [reply httpreq/httpreq.go]: ok,I think so

File: httpreq/httpreq.go (hunk 6 above)
Comment (relax-admin) [reply httpreq/httpreq.go]: I updated this part

File: httpreq/httpreq_test.go (hunk 5 above)
Comment (relax-admin) [reply httpreq/httpreq_test.go]: now,I use this package :"github.com/pangpanglabs/goutils/httpreq"

File: httpreq/httpreq.go [hunk 8]
```diff
@@ -127,11 +164,21 @@ func (r *HttpReq) Call(v interface{}) (int, error) {
...
-		if err := json.Unmarshal(b, &v); err != nil {
+		err = DataTypeFactory{}.New(r.dataType).unMarshal(b, v)
+		if err != nil {
 			return resp.StatusCode, errors.New(string(b))
 		}
 	}
 	return resp.StatusCode, nil
 }
+
+func setDefaultTransport() {
```
Comment (jaehue) [inline httpreq/httpreq.go]: Maybe no where use this function.
If noone use this function, remove it.

File: httpreq/httpreq.go [hunk 9]
```diff
@@ -76,6 +96,18 @@ func (r *HttpReq) WithToken(token string) *HttpReq {
 	return r
 }
 
+func (r *HttpReq) CallWithTransport(transport *http.Transport) *HttpReq {
```
Comment (jaehue) [inline httpreq/httpreq.go]: Do not change `defaultClient`.
Just run `http.Client.Do()` using `transport` parameter.

```
func (r *HttpReq) CallWithTransport(transport *http.Transport, v interface{}) (int, error) {
   // create new *http.Client using transport parameter.
   // and run Client.Do
}
```

Comment (relax-admin) [reply httpreq/httpreq.go]: I modify CallWithTransport method
and added CallWithClient method

File: httpreq/httpreq.go (hunk 8 above)
Comment (relax-admin) [reply httpreq/httpreq.go]: I has deleted it

File: httpreq/httpreq.go (hunk 9 above)
Comment (jaehue) [reply httpreq/httpreq.go]: No. Do not change `defaultClient`.
Only `CallWithTransport` use `transport` parameter.
Others should use default Transport.

If we need change `defaultClient.Transport`, add `SetGlobalTransport` function.

Comment (relax-admin) [reply httpreq/httpreq.go]: I added the method CallWithTransport and SetGlobalTransport

**Comment by jaehue** [comment]
:+1:

**Comment by jaehue** [comment]
:+1:

**Pull request merged**